### PR TITLE
Add support for binaries to run as Windows services

### LINKED ISSUE #59562: Can't run Kubelet as a service on Windows
/kind bug
(this could be considered a feature if this was never intended to work)

**What happened**:

If creating a Windows service through Powershell New-Service command or cmd sc.exe the resulting service cannot start.

**What you expected to happen**:

The service to start

**How to reproduce it (as minimally and precisely as possible)**:

Setup Kubelet.exe on Windows with necessary prereqs. Run in a shell to confirm working. Then take the entire shell command and pass it as the BinaryPathName to New-Service in Powershell. Attempt to start the service through the Services snap-in or with Start-Service.

**Anything else we need to know?**:

I am assuming the signal handler for Kubelet is set up for Linux signals only, and not Windows.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.10.0-alpha.3.299+c17b418f89e056 (manual build)
- Cloud provider or hardware configuration: Local VSphere
- OS (e.g. from /etc/os-release): Windows Server 1709
- Kernel (e.g. `uname -a`):
- Install tools: Manual
- Others:

### PR DESCRIPTION
**Pull request by alinbalutoiu**
**What this PR does / why we need it**:
Add support for binaries to run as Windows services

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #59562

**Special notes for your reviewer**:

**Release note**:
```release-note
kubelet and kube-proxy can now be ran as Windows services
```

### CODE REVIEW

**Comment by aserdean** [comment]
Acked-by: Alin Gabriel Serdean <aserdean@ovn.org>
/assign @bsteciuk @michmike .

**Comment by PatrickLang** [comment]
@jstarks or @jhowardmsft can you take a look since you implemented something similar for dockerd?

File: pkg/util/service/service_windows.go
Comment (PatrickLang) [inline pkg/util/service/service_windows.go]: I think it would be better to bump this up an order of magnitude from 1 to 10 minutes for second retry. I think that would be better for dealing with the case where a dependent service like `dockerd.exe` isn't running yet.

For others reviewing - here's the Windows SDK doc on how restart actions work:
https://msdn.microsoft.com/en-us/library/windows/desktop/ms685939(v=vs.85).aspx 

As written, it will retry after one minute, again after 1 more minute, then wait 24 hours before trying again

**Comment by PatrickLang** [comment]
I added a short comment. Thanks for submitting this and getting restart actions included!

File: pkg/util/service/service_windows.go
Comment (aserdean) [reply pkg/util/service/service_windows.go]: Sure we can change it. I wouldn't go for a hard dependency on `dockerd.exe` since kubelet will be used using a different CRI in the future.

To be honest this is a default value for service creation. The hardcoded values can substitute for a default value, but I would rather add documentation on how to config the service as a sysadmin (i.e. changing start type, adding triggers, modifying restart actions as people pleases). Do you know where we can add such a documentation?

**Review by bsteciuk — APPROVED** [review]
These code changes look good to me.  I also built the bins and verified I was able to register/unregister kube-proxy & kubelet services, as well as do a kubeadm join after setting up the kubelet service (passes kubeadm pre-flight check validating kubelet is setup as a service)

**Comment by cblecker** [comment]
@alinbalutoiu You've added files to vendor without going through the correct process. As such, this will fail CI.
The proper process for adding dependencies is here: https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md

**Comment by alinbalutoiu** [comment]
@cblecker thanks for pointing that out. Done.

**Comment by michmike** [comment]
@alinbalutoiu  please also add the status/approved-for-milestone label

**Comment by alinbalutoiu** [comment]
Fixed the boiler plate for ```service_windows.go```, added missing licenses and added comments to the exported functions from ```pkg/util/service``` package. This were causing the build to fail.

**Comment by michmike** [comment]
/status/approved-for-milestone label

**Comment by alinbalutoiu** [comment]
Resolved the merge conflicts.

**Comment by cblecker** [comment]
@michmike Should this be added to the v1.10 milestone? (I can do that if needed)

**Comment by michmike** [comment]
@cblecker  yes please. we want this in 1.10. thanks!

**Comment by alinbalutoiu** [comment]
Resolved the merge conflict.

**Comment by thockin** [comment]
The issue I have with this is the pkg/util/service.  Service is WAY too overloaded to use as a package name like.  I don't know that the OS-specific hiding should be done at that level.  Instead, what about:

rename it to pkg/windows/service

drop the non-windows parts of it

make kube-proxy and kubelet have an `init_windows.go` file and an `init_others.go` file, with build tags similar to `cmd/kube-proxy/app/server_windows.go` and `cmd/kube-proxy/app/server_others.go`

`init_*.go` defines a function called `initForOS()`

`server.go` calls that.

**Comment by alinbalutoiu** [comment]
Thanks for the review @thockin 
I've addressed all your comments, let me know if this is what you were thinking about.

File: cmd/kube-proxy/app/server.go
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: @mtaufen - aka Mr. Flag.

**Review by thockin — APPROVED** [review]

**Comment by mtaufen** [comment]
/hold
(temporary) I want to take a spin through this to check how flags are registered on the Kubelet, holding since it already has approval and lgtm

File: cmd/kubelet/app/init_windows.go
Comment (mtaufen) [inline cmd/kubelet/app/init_windows.go]: I'd rather have this live in the `options` package with the other flag installers. Maybe in a new file called `k8s.io/kubernetes/cmd/kubelet/app/options/osflags_{windows, other, etc}.go`?

Comment (mtaufen) [reply cmd/kubelet/app/init_windows.go]: e.g. `k8s.io/kubernetes/cmd/kubelet/app/options/osflags_windows.go`:
```
func AddOSFlags(fs *pflag.FlagSet) {
  service.InstallServiceFlags(fs, componentKubelet, "Kubernetes Kubelet")
}
```

File: pkg/windows/service/service.go
Comment (mtaufen) [inline pkg/windows/service/service.go]: You can avoid `*string` in your `var` block at the top of the file, and any associated nil-pointer worries, by registering flags like this:
```
var (
  flServiceName string
  ...
)
flags.StringVar(&flServiceName, "service-name", serviceName, "Set the Windows service name")
...
```

File: pkg/windows/service/service.go
Comment (mtaufen) [inline pkg/windows/service/service.go]: if any of these flags are considered "alpha," please include `<Warning: Alpha feature>` at the beginning of the usage text

File: cmd/kubelet/app/init_windows.go
Comment (mtaufen) [inline cmd/kubelet/app/init_windows.go]: It might be worth using a constant for the service display name.

File: cmd/kubelet/app/init_windows.go
Comment (mtaufen) [reply cmd/kubelet/app/init_windows.go]: I would also look at how the other components in this PR install flags, and try to be consistent with their respective flag-installation patterns.

**Review by mtaufen — CHANGES_REQUESTED** [review]

File: cmd/kubelet/app/init_windows.go
Comment (alinbalutoiu) [reply cmd/kubelet/app/init_windows.go]: Done.

File: pkg/windows/service/service.go
Comment (alinbalutoiu) [reply pkg/windows/service/service.go]: Done.

File: pkg/windows/service/service.go
Comment (alinbalutoiu) [reply pkg/windows/service/service.go]: The flags are not considered "alpha". They are doing a very simple job of registering/removing the kubelet/kube-proxy as Windows services and nothing more.

File: cmd/kubelet/app/init_windows.go
Comment (alinbalutoiu) [reply cmd/kubelet/app/init_windows.go]: Done.

File: cmd/kube-proxy/app/init_windows.go
Comment (mtaufen) [inline cmd/kube-proxy/app/init_windows.go]: @ncdc regarding kube-proxy flag registrations

File: cmd/kube-proxy/app/init_windows.go
Comment (ncdc) [inline cmd/kube-proxy/app/init_windows.go]: This function should return an error instead of calling `glog.Fatal`. Same for the kubelet version.

Comment (ncdc) [reply cmd/kube-proxy/app/init_windows.go]: Would there be some way to make this reusable in a separate package, so we don't have to duplicate it (`initForOS` and `addOSFlags`) in each component?

File: cmd/kubelet/app/server.go
Comment (mtaufen) [inline cmd/kubelet/app/server.go]: Can we call this inside `Run`? `cmd.Run` should only really be loading configuration and passing it to `Run`.

File: cmd/kube-proxy/app/init_windows.go
Comment (ncdc) [reply cmd/kube-proxy/app/init_windows.go]: Shouldn't these go into config instead of being flags?

Comment (mtaufen) [reply cmd/kube-proxy/app/init_windows.go]: I haven't figured out how I want to represent OS-specific configuration parameters in the component config structs yet, so for Kubelet I'd like to keep these as flags for now...

kube-proxy is maybe another story, since we tried to avoid the flag-precedence magic the kubelet implements that allows flags to seamlessly graduate into config fields...

File: cmd/kube-proxy/app/init_windows.go
Comment (alinbalutoiu) [reply cmd/kube-proxy/app/init_windows.go]: Sure, I will update the function to return the error instead of calling ```glog.Fatal```.

In the first version there was a separate package which contained the code. I changed it since @mtaufen wanted this to be inside the ```options```  package from kubelet with the other flag installers. For more details you can view the previous discussion: https://github.com/kubernetes/kubernetes/pull/60144#discussion_r170788021

File: cmd/kubelet/app/server.go
Comment (alinbalutoiu) [reply cmd/kubelet/app/server.go]: I think we can, I will update it.
It doesn't really matter where that it is called, it should be only after the flags are parsed. It just checks if the ```--register-service``` or ```--unregister-flag``` flag was passed in which case it will only create/remove the service and exit. If those flags were not passed, nothing happens.

File: cmd/kube-proxy/app/init_windows.go
Comment (alinbalutoiu) [reply cmd/kube-proxy/app/init_windows.go]: In my opinion this would work only as flags. All the flags are being used only to detect if the user wants to register/remove the kube-proxy as a Windows service.
The flags ```--register-service```/```--unregister-service``` and others added by this should only be called the first time. When the kube-proxy is registered as a Windows service it will not have those flags passed to it, as these are required only the first time to do the Windows registration.
If you would still really like to have those inside the config, I will try to find a way to fit them there.

Comment (ncdc) [reply cmd/kube-proxy/app/init_windows.go]: I'm ok with these being flags. They're similar to kube-proxy's `--cleanup-iptables` in that they're one-shot immediate directives.

Comment (mtaufen) [reply cmd/kube-proxy/app/init_windows.go]: >should only be called the first time

This feels like a strange semantic. You should be able to say "these are the args my thing should run with," and the same set should work for the initial and all subsequent invocations without change - especially for something like kube-proxy, which is specified via a Pod.

As a general recommendation, try to design declarative APIs rather than imperative ones. 

Edit: I guess what you're trying to do is use kube-proxy or kubelet as a bootstrapper for kube-proxy/kubelet as windows services... Is this really what we want to do? 

In kube-proxy's case, don't we lose the visibility we'd typically have via the Pod - not to mention start making lifecycle policy decisions outside of the Kubernetes API?

Kubelet doesn't decide to set itself up as a systemd service, so why should it do what sounds like the same thing for windows? (I know nothing about windows, FYI).

File: pkg/windows/service/service.go
Comment (mtaufen) [inline pkg/windows/service/service.go]: Does `os.Args[0]` here work if the kubelet is started as a hyperkube subcommand?
(does hyperkube support windows?)

File: cmd/kube-proxy/app/init_windows.go
Comment (aserdean) [reply cmd/kube-proxy/app/init_windows.go]: @mtaufen I'm not sure if I follow your logic, but I will try to add more context on what we are trying to achieve.

To break it in more simpler steps `--run-service` is required for the binary to run as a Windows service. Think of having a systemd service but you can only run as a systemd service if and only if you subscribe via an API.

`--register-service`/`--unregister-service` is equivalent to creating/deleting your service via sc.exe (https://support.microsoft.com/ro-ro/help/251192/how-to-create-a-windows-service-by-using-sc-exe for more info) with the caveat that is a commodity for the user. Windows services have a lot of options that can be set: dependencies, restart options with different counters, triggers via events etc. Usually creating/deleting services is done via MSI's, but due to the lack of a Kubernetes MSI we opted to bottle up the experience of creating/deleting the services via those two command line arguments.
`Kubelet doesn't decide to set itself up as a systemd service, so why should it do what sounds like the same thing for windows? `
Nor should it. The user is the one deciding to run the kubelet as a service because the user is the one that passes the command line arguments. Also this is the same approach that has been done on Docker for Windows.
Does the above make sense?

File: pkg/windows/service/service.go
Comment (aserdean) [reply pkg/windows/service/service.go]: To be honest I have no idea if hyperkube is supported on Windows. I don't understand who will try to register the service via hyperkube. It kind of defeats the point of both.

File: cmd/kube-proxy/app/init_windows.go
Comment (mtaufen) [reply cmd/kube-proxy/app/init_windows.go]: >Think of having a systemd service but you can only run as a systemd service if and only if you subscribe via an API.

Is it a hard requirement on Windows that when run as a service a binary must also self-identify as a service? If so, why can't we achieve this with a single `--service` flag?

> Usually creating/deleting services is done via MSI's, but due to the lack of a Kubernetes MSI we opted to bottle up the experience of creating/deleting the services via those two command line arguments.

Ideally, the decision of whether or not to run something as a service would be completely decoupled from the binary and its command-line. If adding these args is just a shortcut because we don't have an MSI, that makes it sound like the right solution is to create an MSI.

> The user is the one deciding to run the kubelet as a service because the user is the one that passes the command line arguments.

I'm not sure I have a clear understanding of the user story here. What are the reasons for running something as a service, and in which cases are these already covered by existing Kubernetes APIs? E.g. I could imagine you want it to be automatically restarted if it crashes, but in the kube-proxy case, why not configure it as a Pod with RestartPolicy=Always?

Comment (aserdean) [reply cmd/kube-proxy/app/init_windows.go]: `Is it a hard requirement on Windows that when run as a service a binary must also self-identify as a service?`
It is a hard requirement. We could use wrappers i.e: https://nssm.cc/usage or https://github.com/cloudbase/OpenStackService, but you gain more advantages when doing from the actual binary.
`If so, why can't we achieve this with a single --service flag?`
We could. Although we prefer possible to have all three options for the moment. We shall remove this step when we can bundle the kubernetes installation via MSI.
`I'm not sure I have a clear understanding of the user story here. What are the reasons for running something as a service, and in which cases are these already covered by existing Kubernetes APIs? E.g. I could imagine you want it to be automatically restarted if it crashes, but in the kube-proxy case, why not configure it as a Pod with RestartPolicy=Always?`
That would be just one feature that you can access.
There are lot of hidden dragons that you can unlock. I.E. Running as an elevated user i.e. LocalSytstem (https://msdn.microsoft.com/en-us/library/windows/desktop/ms684190(v=vs.85).aspx).
You can add triggers for first/second/rest of the restarts to trigger a script(cleanup/dumps/etc) and reboot the node if required.
Instead of starting blindly you can wait for dependencies: i.e. I don't have docker or kubelet running, I add docker as a dependency to the kubelet service. Whenever I start the kubelet service I also start docker. Service manager is the one that starts docker and after starts kubelet. 
I could add more examples if needed.

Comment (mtaufen) [reply cmd/kube-proxy/app/init_windows.go]: >It is a hard requirement.

Am I understanding right that even with the MSI we'll still need something like `--service`? In that case, ok then. Can't fight the OS APIs. 

>We shall remove this step when we can bundle the kubernetes installation via MSI.

How long will it take to create an MSI/what is the risk of *not* having these flags in this release? I'd rather not even have to carry the temporary bits if possible.

File: pkg/windows/service/service.go
Comment (PatrickLang) [reply pkg/windows/service/service.go]: I'm not aware of any hyperkube support for Windows, and running the kubelet in a container won't work because `--privileged` isn't available on Windows. It needs to be run on the host directly.

File: cmd/kube-proxy/app/init_windows.go
Comment (PatrickLang) [reply cmd/kube-proxy/app/init_windows.go]: I don't think we should count on MSI as a means for configuration management. It hasn't received any new features in years. The guidance for desktop developers since Windows 8 has been to move to APPX but there's no equivalent for servers. This is the same reasoning that there isn't a MSI installer for Docker EE-basic on Windows Server.

I think this solution with the `--register-service` flag is the right one and will be easier to use with configuration managers such as PowerShell Desired State Configuration, Ansible, Chef, Puppet, and so on. MSI is not idempotent, and that's a property that configuration managers prefer to have.

Comment (mtaufen) [reply cmd/kube-proxy/app/init_windows.go]: >MSI is not idempotent, and that's a property that configuration managers prefer to have.

These are not idempotent flags, because they have to be stripped from the command line before registering the component as a service.

>use with configuration managers such as PowerShell Desired State Configuration, Ansible, Chef, Puppet, and so on

are *none* of these capable of configuring something to run as a service?

Comment (PatrickLang) [reply cmd/kube-proxy/app/init_windows.go]: They could shell out to `sc.exe` to register a service. 

My main point was if you ran a playbook that called `kubelet.exe --register-service ...` it would do the right thing. If you modified your playbook with different flags after that, the new settings from the next `kubelet.exe --register-service` would still apply correctly. `--unregister-service` would always work as well. 

If you don't want functionality that makes it easy to configure the service, it could be removed and handled as extra steps in an outside system but `--service` is absolutely essential to enable the Windows service control manager API integration.

Comment (aserdean) [reply cmd/kube-proxy/app/init_windows.go]: `Am I understanding right that even with the MSI we'll still need something like --service? In that case, ok then. Can't fight the OS APIs.`
Yup `--service` is mandatory. It's OS specific.
`How long will it take to create an MSI/what is the risk of not having these flags in this release? I'd rather not even have to carry the temporary bits if possible.`
I think the next release cycle might have support we might add support for it (if we find the time for it).
@mtaufen the main risk here if we drop `--register-service`/`--unregister-service` is that we have to at least document it somewhere, but I'm not exactly sure where and if the documentation changes can land into 1.10 milestone.
How should we proceed?

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60144)

File: cmd/kube-proxy/app/init_windows.go
Comment (mtaufen) [reply cmd/kube-proxy/app/init_windows.go]: Can you point me to links where documentation would need to change?
We didn't have `--register-service`, etc. before this PR, so I'm not sure why there would be documentation indicating otherwise. 
Generally, documentation lives in https://github.com/kubernetes/website.

Regarding `--service`, is there any way we could have the Windows build of a Kubelet automatically detect that someone is trying to configure it as a service, and avoid needing any flags at all?

Comment (aserdean) [reply cmd/kube-proxy/app/init_windows.go]: @mtaufen unfortunately `--service` is `mandatory`.

We are dropping the `--register-service`/`--unregister-service` part from this PR.

**Comment by alinbalutoiu** [comment]
@mtaufen I updated the PR, we dropped --register-service and --unregister-service and removed all the related code for the flags.
``--service`` is mandatory unfortunately.

File: pkg/windows/service/service.go
Comment (mtaufen) [reply pkg/windows/service/service.go]: @deads2k do you know?

Comment (deads2k) [reply pkg/windows/service/service.go]: > @deads2k do you know?

I do not know.

File: cmd/kube-proxy/app/server.go
Comment (mtaufen) [inline cmd/kube-proxy/app/server.go]: If dropping the register/unregister flags, maybe this error message should change as well?

File: cmd/kube-proxy/app/init_windows.go
Comment (mtaufen) [inline cmd/kube-proxy/app/init_windows.go]: suggest:
```
if runAsService, err := service.InitService(serviceName); err != nil {
  return err
} else if runAsService {
  glog.Infof("Running %s as a Windows service!", serviceName)
}
```

Comment (mtaufen) [reply cmd/kube-proxy/app/init_windows.go]: Alternatively, could the log be written in `service.InitService`?

File: cmd/kube-proxy/app/server.go
Comment (mtaufen) [inline cmd/kube-proxy/app/server.go]: Is a constructor really the best place to call this? I'd expect to see init calls like this in a bootstrapping function, like Run. @ncdc

File: cmd/kubelet/app/init_windows.go
Comment (mtaufen) [inline cmd/kubelet/app/init_windows.go]: similar suggestions here to kube-proxy's copy of this code, above

File: cmd/kubelet/app/server.go
Comment (mtaufen) [inline cmd/kubelet/app/server.go]: Not sure this comment is necessary; I could see `initForOS` growing to do other OS-specific non-Windows things in the future.

File: pkg/windows/service/service.go
Comment (mtaufen) [inline pkg/windows/service/service.go]: nit: delete this empty line

File: pkg/windows/service/service.go
Comment (mtaufen) [inline pkg/windows/service/service.go]: Since this flag is Windows-specific, we should have a name that indicates as much.
Some potential options:
- `windows-service`
- `enable-windows-scm`

File: pkg/windows/service/service.go
Comment (mtaufen) [inline pkg/windows/service/service.go]: update comment to reflect that registration functionality is gone

File: pkg/windows/service/service.go
Comment (mtaufen) [inline pkg/windows/service/service.go]: I also think it might be better to push the ownership of the flag and destination value to the components, rather than this library. (E.g. component owns `--service` and an analogue to `flRunService`, and chooses to call `InitService` if the flag is true).

For example, if we want to move `--service` into the Kubelet's config file API in the future, it's much easier if it's just a field move from `cmd/kubelet/app/options.KubeletFlags` to `pkg/kubelet/apis/kubeletconfig/v1beta1.KubeletConfiguration` and an update to the `flags.BoolVar` call. If we don't do this, we'll probably have to write custom logic to plumb the value from the config file through the flag API to reach `flRunService`, which will may make it hard to ever deprecate the flag.

File: pkg/windows/service/service.go
Comment (mtaufen) [inline pkg/windows/service/service.go]: see below comment about components just choosing to call `InitService` based on a flag that they own.

**Comment by mtaufen** [comment]
@alinbalutoiu thanks, I made another review pass
Let's also change the PR title/description and commit text to reflect that this is now just adds a Windows SCM integration, and not a service registration utility.

File: cmd/kube-proxy/app/server.go
Comment (ncdc) [reply cmd/kube-proxy/app/server.go]: @mtaufen this is part of `Run`. When I first looked at your comment, it looked like GH was showing it as part of `NewOptions` (so my reaction was in line with your comment), but if you expand the lines above this diff, you'll see it's in `Run`.... right?

Comment (mtaufen) [reply cmd/kube-proxy/app/server.go]: Oh GitHub xD. You're right.

File: cmd/kube-proxy/app/init_windows.go
Comment (mtaufen) [reply cmd/kube-proxy/app/init_windows.go]: >unfortunately --service is mandatory

Oh well. Ok then, that'll have to do.

>We are dropping the --register-service/--unregister-service part from this PR.

I'm confused. Do existing docs need to be updated to reflect this, or did you just want to add documentation that clearly indicates what the Kubelet is/isn't capable of wrt Windows SCM?

**Comment by PatrickLang** [comment]
Alright - think labels are fixed. Just need final review closed from @mtaufen I hope?

File: cmd/kube-proxy/app/init_windows.go
Comment (mtaufen) [inline cmd/kube-proxy/app/init_windows.go]: @ncdc this global state look good? Or does kube-proxy prefer something akin to `KubeletFlags`?

File: cmd/kubelet/app/options/osflags_windows.go
Comment (mtaufen) [inline cmd/kubelet/app/options/osflags_windows.go]: This should be a subfield of the `KubeletFlags` struct in `options/options.go`. Leave a comment above the field noting that its corresponding flag only gets registered in Windows builds.

Let's also rename it to `WindowsService` here and in kube-proxy so it matches the flag name.

File: cmd/kubelet/app/options/osflags_windows.go
Comment (mtaufen) [inline cmd/kubelet/app/options/osflags_windows.go]: Modify this as follows when you move the destination value into `KubeletFlags`, so we're consistent with `KubeletFlags.AddFlags`:
```
func (f *KubeletFlags) addOSFlags(fs *pflag.FlagSet) {
  fs.BoolVar(&f.WindowsService, "windows-service", f.WindowsService, "Enable Windows Service Control Manager API integration")
}
```
and now call it at the beginning of `KubeletFlags.AddFlags` instead of in `cmd/kubelet/app/server.go` (I made `KubeletFlags.addOSFlags` private, since there shouldn't be a reason to call this separately from `KubeletFlags.AddFlags`).

File: cmd/kubelet/app/server.go
Comment (mtaufen) [inline cmd/kubelet/app/server.go]: merge with kubeletFlags.AddFlags, per above comment

File: cmd/kubelet/app/init_windows.go
Comment (mtaufen) [inline cmd/kubelet/app/init_windows.go]: When you remove the global `FlRunService` var in favor of `KubeletFlags.WindowsService`, you'll have to pass it in as an arg here, e.g. `func initForOS(windowsService bool) error`.

I know that's not as pretty a function signature, but it's better than global state floating around and keeps things consistent with the other flags.

**Review by mtaufen — COMMENTED** [review]
Updates mostly look good, I had a few more suggestions to tighten things up.

File: cmd/kube-proxy/app/init_windows.go
Comment (ncdc) [reply cmd/kube-proxy/app/init_windows.go]: I had put everything into the `Options` struct, such as `CleanupAndExit`:

https://github.com/kubernetes/kubernetes/blob/86ca3afb4d03c769d4e643eb0320f17cdf7f9903/cmd/kube-proxy/app/server.go#L93-L100

And then the flag registered directly into that field:

https://github.com/kubernetes/kubernetes/blob/86ca3afb4d03c769d4e643eb0320f17cdf7f9903/cmd/kube-proxy/app/server.go#L120-L124

But I realize this is OS specific and it doesn't make sense to add it to kube-proxy's Options struct, so it's probably ok like this.

FWIW, I wouldn't prefix the package variable's name with `fl`. I'd probably call it `runAsWIndowsServer` or something to that effect instead.

Comment (alinbalutoiu) [reply cmd/kube-proxy/app/init_windows.go]: Done. I changed the variable name into ``WindowsService`` to be consistent with Kubelet as @mtaufen requested.

File: cmd/kubelet/app/options/osflags_windows.go
Comment (alinbalutoiu) [reply cmd/kubelet/app/options/osflags_windows.go]: Done

File: cmd/kubelet/app/options/osflags_windows.go
Comment (alinbalutoiu) [reply cmd/kubelet/app/options/osflags_windows.go]: Done

File: cmd/kubelet/app/server.go
Comment (alinbalutoiu) [reply cmd/kubelet/app/server.go]: Done

File: cmd/kubelet/app/init_windows.go
Comment (alinbalutoiu) [reply cmd/kubelet/app/init_windows.go]: Done

File: cmd/kube-proxy/app/server.go
Comment (mtaufen) [inline cmd/kube-proxy/app/server.go]: s/kubelet/kube-proxy

**Review by mtaufen — COMMENTED** [review]
One nit, and also want @ncdc to give a final thumbs-up on the kube-proxy stuff, otherwise LGTM.

File: cmd/kube-proxy/app/server.go
Comment (alinbalutoiu) [reply cmd/kube-proxy/app/server.go]: Done

**Review by mtaufen — APPROVED** [review]

**Comment by mtaufen** [comment]
I pinged @ncdc last night on Slack, and he was cool with the kube-proxy side.
Kubelet side LGTM.
/lgtm

**Pull request merged**
